Add editorconfig configuration

faern opened the pull request:
We have lately had some formatting problems in some PRs. Mainly it has been tabs mixed in with spaces for indentation as well as [BOM](https://en.wikipedia.org/wiki/Byte_order_mark) bytes at the start of some files saved in Visual Studio.

I have discussed this with Odd and Emils. Apparently there is no easy way to make VS not do this. And according to what I have heard the "standardized" [EditorConfig](https://editorconfig.org/#overview) should be able to fix this. EditorConfig is also supported by Vim, Visual Studio Code etc, but only with a plugin. You can see the full list of supported editors on their website. But as long as your editor does not break these rules you gain nothing from supporting this configuration file anyway.

So this file does not change any format we have in this repo, it just restates them in a shared config in the hopes of reducing the risk of commiting files with undesired formats.

My goal was not to be as verbose as possible in the config. Rather keeping it minimal. Only stating the configuration we have been having problems with. Feel free to comment on the stuff I added so far and suggest more additions if you feel your editing experience could be improved by another setting.

I'm usually not a fan of having editor configs checked in. But this will help us to get started when moving to new VMs and computers, something we do fairly often when testing on all the platforms.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)
<!-- Reviewable:end -->

faern commented:
Review status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @mvd-ows, @jvff, @pronebird, and @pinkisemils)

---

*[.editorconfig, line 7 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BrlwBQT-0_yfIiYv:-LG-BrlwBQT-0_yfIiYw:bgxrsyi) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L7)):*
> ```INI
> 
> [*]
> charset = utf-8
> ```

This is a bit cryptic. But according to rumor, setting this should stop VS from adding BOM bytes to files(?)

---

*[.editorconfig, line 8 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BwaJ1np0FeeP-VJr:-LG-BwaJ1np0FeeP-VJs:b-svprbe) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L8)):*
> ```INI
> [*]
> charset = utf-8
> end_of_line = lf
> ```

Will this work for Windows users? I thought that maybe it would make it possible to work solely in LF. If you think this will cause more problems than it will solve we will of course not have this setting.

I was thinking that since there has been some problems with rustfmt in regards to line endings this might help?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

mvd-ows commented:
*[.editorconfig, line 8 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BwaJ1np0FeeP-VJr:-LG-G_C96jppMy29pMSr:b-jx709k) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L8)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Will this work for Windows users? I thought that maybe it would make it possible to work solely in LF. If you think this will cause more problems than it will solve we will of course not have this setting.

I was thinking that since there has been some problems with rustfmt in regards to line endings this might help?
</blockquote></details>

I don't think VS cares about what type of line break the file is using. The default on Windows and in VS is to use CRLF. All C++ code we currently have seems to be using CRLF, or perhaps it's `git` rewriting the line breaks on checkout.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

mvd-ows commented:
*[.editorconfig, line 7 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BrlwBQT-0_yfIiYv:-LG-HRT-Dlcf81wzNtao:bm9cmnw) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L7)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This is a bit cryptic. But according to rumor, setting this should stop VS from adding BOM bytes to files(?)
</blockquote></details>

OK, nice.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

mvd-ows commented:
Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @jvff, @pronebird, @pinkisemils, @mvd-ows, and @faern)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @jvff, @pinkisemils, @mvd-ows, @pronebird, and @faern)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

mvd-ows commented:
*[.editorconfig, line 8 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BwaJ1np0FeeP-VJr:-LG-IQIYDg07zsXThgyN:b8j2zfa) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L8)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

I don't think VS cares about what type of line break the file is using. The default on Windows and in VS is to use CRLF. All C++ code we currently have seems to be using CRLF, or perhaps it's `git` rewriting the line breaks on checkout.
</blockquote></details>

There's another PR that updates `rustfmt.toml` with `newline_style = "Native"`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @jvff, @pinkisemils, @mvd-ows, @pronebird, and @faern)

---

*[.editorconfig, line 7 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BrlwBQT-0_yfIiYv:-LG-IpTY0oTivp4xIT1I:b-dh5hos) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L7)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

OK, nice.
</blockquote></details>

I hope so. @pinkisemils told me this :)

---

*[.editorconfig, line 8 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BwaJ1np0FeeP-VJr:-LG-Iczw6bL56qUU6O_T:bk5db7y) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L8)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

There's another PR that updates `rustfmt.toml` with `newline_style = "Native"`.
</blockquote></details>

Yes. And this addition is because of that PR. We should only need one of them. If we merge this we probably don't need `newline_style = "Native"`. But that has to be tested by someone using Windows I think.

So the thing is. Either we go full LF on Windows (can be restricted to just Rust files if we want to limit the scope) and then rustfmt can work in its default LF mode and also git does not have to convert line endings on commit/checkout.

Or we continue using CRLF on Windows, then git will have to do the conversion, because commited to git we should stick to one type of line ending, and that is currently LF. Problem with this is that there seems to be some bugs with CRLF for rustfmt, so I suggested this as a remedy, but I don't know if it will solve it or not.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @jvff, @pinkisemils, @mvd-ows, @pronebird, and @faern)

---

*[.editorconfig, line 8 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BwaJ1np0FeeP-VJr:-LG-LjuP-e7VewORynZk:b-8gxfur) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L8)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Sure. If you think git or other tools won't go nuts if we have mixed line ending styles in the same repo?
</blockquote></details>

(We can try, it's not like we can't change this file later)

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @jvff, @pinkisemils, @mvd-ows, @pronebird, and @faern)

---

*[.editorconfig, line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-KNRH3Ci-w2Kpb8_L:-LG-KNRH3Ci-w2Kpb8_M:b-5nfjfi) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L16)):*
> ```INI
> indent_size = 4
> 
> [*.js]
> ```

`prettier` reads `.editorconfig`, so we should either figure out how to integrate both and maybe drop duplicate rules from `.prettierrc`. Also, we have other files such as `.css` and `.html` and `.json` besides `.js`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @jvff, @pinkisemils, @mvd-ows, @pronebird, and @faern)

---

*[.editorconfig, line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-KNRH3Ci-w2Kpb8_L:-LG-NO0bCOi6ziD5ZCSq:b2br38d) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L16)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

`prettier` reads `.editorconfig`, so we should either figure out how to integrate both and maybe drop duplicate rules from `.prettierrc`. Also, we have other files such as `.css` and `.html` and `.json` besides `.js`.
</blockquote></details>

ps: prettier supports the following configuration fields:

```
indent_style
indent_size/tab_width
max_line_length
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 2 files at r2.
Review status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @mvd-ows, @jvff, @pronebird, @pinkisemils, and @faern)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

mvd-ows commented:
Reviewed 1 of 2 files at r2.
Review status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @mvd-ows, @jvff, @pronebird, @pinkisemils, and @faern)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

mvd-ows commented:
Reviewed 1 of 2 files at r2.
Review status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jvff, @pronebird, @pinkisemils, @mvd-ows, and @faern)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pinkisemils commented:
Review status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jvff, @pronebird, @pinkisemils, @mvd-ows, and @faern)

---

*[.editorconfig, line 7 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-BrlwBQT-0_yfIiYv:-LG-OABs6tJNyjphMLGh:b-8pau7p) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L7)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I hope so. @pinkisemils told me this :)
</blockquote></details>

Seems to be the case for new files.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @mvd-ows, @jvff, @pronebird, @pinkisemils, and @faern)

---

*[.editorconfig, line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-KNRH3Ci-w2Kpb8_L:-LG-Qn3Q0WpUZyeageHl:b-nhe35o) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L16)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

ps: prettier supports the following configuration fields:

```
indent_style
indent_size/tab_width
max_line_length
```
</blockquote></details>

Alright cool, leave it as is for now. I spotted that I only format `*.js` files anyway and we're going to get rid of almost all of *.css` files anyway soon.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @mvd-ows, @jvff, @pronebird, @pinkisemils, and @faern)

---

*[.editorconfig, line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-KNRH3Ci-w2Kpb8_L:-LG-QtFSAj9ACEmNlpKq:b-p6uovl) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L16)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

Alright cool, leave it as is for now. I spotted that I only format `*.js` files anyway and we're going to get rid of almost all of *.css` files anyway soon.
</blockquote></details>

Added css/html/json and removed the equivalent settings from `prettier`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 2 files at r3.
Review status: 0 of 3 files reviewed, all discussions resolved (waiting on @mvd-ows, @jvff, @pronebird, and @pinkisemils)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 3 files reviewed, all discussions resolved (waiting on @mvd-ows, @jvff, @pronebird, and @pinkisemils)

---

*[.editorconfig, line 16 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-KNRH3Ci-w2Kpb8_L:-LG-R7ED95uStWnBJ8Yi:bhzvliq) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/516a628bacbc11bbd8863c44079d10d63c343c75/.editorconfig#L16)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Added css/html/json and removed the equivalent settings from `prettier`.
</blockquote></details>

Great :+1:

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @mvd-ows, @jvff, @pronebird, @pinkisemils, and @faern)

---

*[.editorconfig, line 8 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-RDsC27jmNoOdTSjR:-LG-RDsC27jmNoOdTSjS:b-vdtj4x) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a0a02c4594950818c37dd67803b7f70393d6c364/.editorconfig#L8)):*
> ```INI
> [*]
> charset = utf-8
> insert_final_newline = true
> ```

final newline may break "JSON".

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 3 files reviewed, all discussions resolved (waiting on @mvd-ows, @jvff, @pronebird, and @pinkisemils)

---

*[.editorconfig, line 8 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-LG-RDsC27jmNoOdTSjR:-LG-RzkKAZYD75AOwJIX:b-e4uwgc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a0a02c4594950818c37dd67803b7f70393d6c364/.editorconfig#L8)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

final newline may break "JSON".
</blockquote></details>

Just checked on Webkit and it parses JSON regardless trailing newlines.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 2 files at r3.
Review status: 0 of 3 files reviewed, all discussions resolved (waiting on @mvd-ows, @jvff, @pinkisemils, and @pronebird)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

pronebird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed, all discussions resolved (waiting on @mvd-ows, @jvff, @pinkisemils, and @pronebird)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-:-LG-SFw_5ZAFELEZ2rLO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pinkisemils commented:
Reviewed 1 of 1 files at r1, 1 of 2 files at r2, 1 of 2 files at r3.
Review status: 0 of 3 files reviewed, all discussions resolved (waiting on @jvff, @pinkisemils, @mvd-ows, and @pronebird)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248)*
<!-- Sent from Reviewable.io -->

jvff commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 2 of 2 files at r3.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/248#-:-LG-ugrj0x31hJKJ6ZZx:bnfp4nl)*
<!-- Sent from Reviewable.io -->
